File: humanizer-engine/python-backup/ninja_post_processor.py
import re
import random

def phase_1_structural_cleanup(text: str) -> str:
    """Fix sentence boundaries, double spaces, hanging punctuation."""
    text = re.sub(r'\s+', ' ', text)
    text = re.sub(r'([.?!])\s*([a-z])', lambda m: f"{m.group(1)} {m.group(2).upper()}", text)
    text = re.sub(r'\s+([,.:;?!])', r'\1', text)
    text = text.replace('.. ', '. ')
    text = text.replace('..', '.')
    return text.strip()

def phase_2_academic_style(text: str) -> str:
    """Balance transition words, remove cliches."""
    fluff_words = ['game-changing', 'crucial', 'leveraging', 'in conclusion', 'to summarize']
    for word in fluff_words:
        text = re.sub(rf'\b{word}\b', '', text, flags=re.IGNORECASE)
    
    # Simple transition replacements
    text = text.replace('Furthermore,', 'In addition,').replace('Moreover,', 'Additionally,')
    return re.sub(r'\s+', ' ', text).strip()

def phase_3_clarity_compression(text: str) -> str:
    """Trim repetitive phrases locally."""
    text = text.replace('due to the fact that', 'because')
    text = text.replace('in order to', 'to')
    text = text.replace('a large number of', 'many')
    return text

def phase_4_fact_and_format(text: str) -> str:
    """Ensure grammatical consistency and no orphaned bullets."""
    text = re.sub(r'^- ', '', text, flags=re.MULTILINE)  # clear basic bullets if any snuck in
    text = text.replace(' ,', ',').replace(' .', '.')
    return text

def execute_ninja_non_llm_phases(text: str) -> str:
    """Run all 4 non-LLM post-processing phases."""
    text = phase_1_structural_cleanup(text)
    text = phase_2_academic_style(text)
    text = phase_3_clarity_compression(text)
    text = phase_4_fact_and_format(text)
    return text

File: humanizer-engine/frontend/lib/engine/ninja-post-processor.ts
/**
 * Ninja Post-Processor — Deterministic Non-LLM Phases
 * ====================================================
 * 4 quick cleanup passes run after the LLM pipeline.
 */

/** Phase 1: Fix sentence boundaries, double spaces, hanging punctuation. */
function phase1StructuralCleanup(text: string): string {
  text = text.replace(/\s+/g, " ");
  text = text.replace(/([.?!])\s*([a-z])/g, (_, p, c) => `${p} ${c.toUpperCase()}`);
  text = text.replace(/\s+([,.:;?!])/g, "$1");
  text = text.replace(/\.\. /g, ". ").replace(/\.\./g, ".");
  return text.trim();
}

/** Phase 2: Balance transition words, remove clichés. */
function phase2AcademicStyle(text: string): string {
  const fluff = ["game-changing", "crucial", "leveraging", "in conclusion", "to summarize"];
  for (const word of fluff) {
    text = text.replace(new RegExp(`\\b${word.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}\\b`, "gi"), "");
  }
  text = text.replace(/Furthermore,/g, "In addition,").replace(/Moreover,/g, "Additionally,");
  return text.replace(/\s+/g, " ").trim();
}

/** Phase 3: Trim repetitive phrases locally. */
function phase3ClarityCompression(text: string): string {
  text = text.replace(/due to the fact that/gi, "because");
  text = text.replace(/in order to/gi, "to");
  text = text.replace(/a large number of/gi, "many");
  return text;
}

/** Phase 4: Ensure grammatical consistency, no orphaned bullets. */
function phase4FactAndFormat(text: string): string {
  text = text.replace(/^- /gm, "");
  text = text.replace(/ ,/g, ",").replace(/ \./g, ".");
  return text;
}

/** Run all 4 non-LLM post-processing phases. */
export function executeNinjaNonLlmPhases(text: string): string {
  text = phase1StructuralCleanup(text);
  text = phase2AcademicStyle(text);
  text = phase3ClarityCompression(text);
  text = phase4FactAndFormat(text);
  return text;
}
